C++ mode generates bad initialization code for union type

### Description

```nim
type
  Union{.union.} = object
    a:int32
    b:bool
const i = Union(b:true)
echo i
```
C++ backend generates incorrect initialization code for this ( see below ).
C backend works fine.


### Nim Version

Nim Compiler Version 2.1.1 [MacOSX: arm64]
Compiled at 2023-09-12
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 8f5b90f886501862bd27d4e0e8244e2f7f0ebc2a
active boot switches: -d:release -d:danger

### Current Output

```text
.................................................................../Volumes/prj/pil/test/cpptest/_build/debug/.nimcache/@munionbug.nim.cpp:60:93: error: excess elements in union initializer
N_LIB_PRIVATE NIM_CONST tyObject_Union__TAj7qkZEjaueQFP7zLlauw i__unionbug_u4 = {((NI32)0), NIM_TRUE}
                                                                                            ^~~~~~~~
/Volumes/prj/dev/nimdev/lib/nimbase.h:320:18: note: expanded from macro 'NIM_TRUE'
#define NIM_TRUE true
                 ^~~~
1 error generated.

Error: execution of an external compiler program 'clang++ -c -std=gnu++17 -funsigned-char  -pthread -g   -I/Volumes/prj/dev/nimdev/lib -I/Volumes/prj/pil/test/cpptest -o /Volumes/prj/pil/test/cpptest/_build/debug/.nimcache/@munionbug.nim.cpp.o /Volumes/prj/pil/test/cpptest/_build/debug/.nimcache/@munionbug.nim.cpp' failed with exit code: 1
```


### Expected Output

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

--- a/compiler/vmgen.nim
+++ b/compiler/vmgen.nim
@@ -2050,6 +2050,8 @@ proc genSetConstr(c: PCtx, n: PNode, dest: var TDest) =
       c.freeTemp(a)
 
 proc genObjConstr(c: PCtx, n: PNode, dest: var TDest) =
+  if tfUnion in n.typ.flags: # bug #22708 # bug #13481
+    globalError(c.config, n.info, "object with '{.union.}' pragmas is not supported by VM")
   if dest < 0: dest = c.getTemp(n.typ)
   let t = n.typ.skipTypes(abstractRange+{tyOwned}-{tyTypeDesc})
   if t.kind == tyRef:

--- a/tests/vm/tvmmisc.nim
+++ b/tests/vm/tvmmisc.nim
@@ -301,14 +301,6 @@ block: # bug #10815
   const a = P()
   doAssert $a == ""
 
-when defined osx: # xxx bug #13481
-  block:
-    type CharSet {.union.} = object
-      cs: set[char]
-      vs: array[4, uint64]
-    const a = Charset(cs: {'a'..'z'})
-    doAssert a.repr.len > 0
-
 import tables
 
 block: # bug #8007

